dipy/core/gradients.py

```diff
@@ -1343,3 +1343,133 @@ def _check_ornt(ornt):
     if tuple(uniq) not in {(-1, 1), (-1,), (1,)}:
         print(tuple(uniq))
         return True
+
+
+def extract_b0(dwi, b0_mask, *, group_contiguous_b0=False, strategy="mean"):
+    """Extract a set of b0 volumes from a dwi dataset.
+
+    Parameters
+    ----------
+    dwi: ndarray
+        4D diffusion-weighted imaging data (X, Y, Z, N), where N is the number of
+        volumes.
+    b0_mask: ndarray of bool
+        Mask over the time dimension (4th) identifying b0 volumes.
+    group_contiguous_b0: bool, optional
+        If True, each contiguous b0 volumes are grouped together.
+    strategy: str, optional
+        The extraction strategy, of either:
+
+            - first: select the first b0 found.
+            - all: select them all.
+            - mean: average them.
+
+        When used in conjunction with the batch parameter set to True, the strategy
+        is applied individually on each continuous set found.
+
+    Returns
+    -------
+    b0_data : ndarray
+        Extracted b0 volumes.
+    """
+    if not isinstance(b0_mask, np.ndarray) or b0_mask.ndim != 1:
+        raise ValueError("b0_mask must be a boolean array.")
+
+    if not isinstance(dwi, np.ndarray) or dwi.ndim != 4:
+        raise ValueError("dwi must be a 4D numpy array.")
+
+    if dwi.shape[-1] != b0_mask.shape[0]:
+        raise ValueError("The last dimension of dwi must match the length of b0_mask.")
+
+    strategy = strategy.lower()
+    if strategy not in ["first", "all", "mean"]:
+        raise ValueError(
+            "Invalid strategy: {}. Valid strategies are: " "first, all, mean.".format(
+                strategy
+            )
+        )
+
+    if group_contiguous_b0:
+        b0_indices = np.where(b0_mask)[0]
+        groups = np.split(b0_indices, np.where(np.diff(b0_indices) > 1)[0] + 1)
+        grouped_b0 = []
+        for group in groups:
+            group_data = dwi[..., group - b0_indices[0]]
+            if strategy == "first":
+                grouped_b0.append(group_data[..., 0])
+            elif strategy == "mean":
+                grouped_b0.append(np.mean(group_data, axis=-1))
+            else:  # "all"
```

Comment:
The behavior of "all" here confuses me. This would return a 5D array, which does not seem ideal. Shouldn't it concatenate instead of stack in this case?

Reply:
I understand then it would be the same behavior as using "all" without `group_contiguous_b0`. We could even then just change the code to check if strategy is "all" and return `b0_data` in that case regardless of what `group_contiguous_b0`

Reply:
@pjsjongsung If we do that we will be loosing the grouping information of the b0s. Having said that we also need to tell the user explicitly as the use-case of `strategy="all"` with `group_contiguous_bo=True` is not possible as it is an inconsistent behavior from what is happening with other strategies.

Reply:
Agree with @maharshi-gor here. let me know if you want more detailed explanation.